src/main/scala/sbtcrossproject/CrossProject.scala

```diff
@@ -1,184 +1,119 @@
-      crossType: CrossType): CrossProject = {
 
-    val sss = sharedSrcSettings(crossType)
-
-    val projects = ids.map { case (p, id) =>
-      val project = Project(id, crossType.dir(p, base)).
-        settings(sss: _*).
-        enablePlugins(p.plugin.toSeq: _*)
-
-      p -> project
+  def crossProject_impl[A <: Platform](c: Context): c.Expr[Builder[A]] = {
+    import c.universe._
+    val enclosingValName = MacroUtils.definingValName(
+        c,
+        methodName =>
+          s"""$methodName must be directly assigned to a val, such as `val x = $methodName`.""")
+    val name = c.Expr[String](Literal(Constant(enclosingValName)))
+    reify {
+      new Builder(name.splice, file("."), Map.empty)
     }
-
-    new CrossProject(crossType, projects)
   }
 
-  private def sharedSrcSettings(crossType: CrossType) = Seq(
-      unmanagedSourceDirectories in Compile ++= {
-        makeCrossSources(crossType.sharedSrcDir(baseDirectory.value, "main"),
-            scalaBinaryVersion.value, crossPaths.value)
-      },
-      unmanagedSourceDirectories in Test ++= {
-        makeCrossSources(crossType.sharedSrcDir(baseDirectory.value, "test"),
-            scalaBinaryVersion.value, crossPaths.value)
-      }
-  )
-
-  // Inspired by sbt's Defaults.makeCrossSources
-  private def makeCrossSources(sharedSrcDir: Option[File],
-      scalaBinaryVersion: String, cross: Boolean): Seq[File] = {
-    sharedSrcDir.fold[Seq[File]] {
-      Seq.empty
-    } { srcDir =>
-      if (cross)
-        Seq(srcDir.getParentFile / s"${srcDir.name}-$scalaBinaryVersion", srcDir)
-      else
-        Seq(srcDir)
-    }
-  }
+  implicit def toCrossProject[A <: Platform](
+      builder: Builder[A]): CrossProject[A] =
+    new CrossProject[A](builder.id, builder.dir, builder.map)
 
-  final class Builder(id: String, platforms: Seq[CrossPlatform], base: File) {
-    def crossType(crossType: CrossType): CrossProject =
-      CrossProject(id, platforms, base, crossType)
-  }
+  final class Builder[A <: Platform](val id: String,
+                                     val dir: File,
+                                     val map: Map[Platform, Project]) {
 
-  def crossProject_impl(c: Context)(platforms: c.Expr[CrossPlatform]*): c.Expr[Builder] = {
-    import c.universe._
-    val enclosingValName = MacroUtils.definingValName(c, methodName =>
-      s"""$methodName must be directly assigned to a val, such as `val x = $methodName`.""")
-    val name = c.Expr[String](Literal(Constant(enclosingValName)))
-    val SeqApplyPath = Seq("scala", "collection", "Seq", "apply")
-    val SeqApply = SeqApplyPath.foldLeft[Tree](Ident("_root_")) {
-      case (t, name) => Select(t, newTermName(name))
+    def in(dir: File): Builder[A] = {
+      new Builder[A](id, dir, map)
+    }
+
+    def platform[U <: Platform](t: U): Builder[A with U] = {
+      new Builder[A with U](id,
+                            dir,
+                            map + (t -> Project(t.name, file(t.path))))
```

Comment:
This doesn't take `dir` into account in the `file` part of the project.

I suggest that `Builder` only keeps a `List[Platform]` instead of the constructed projects. Then the conversion to `CrossProject` can map over that list to create the `Project`s, making sure to give `dir + t.path` to each project.

If you use `dir` right now to create the `Project`, there will be confusing effects of the user calls `.platform` before `.in`.